bullet-featherstone/src/Base.cc

```diff
@@ -26,14 +26,213 @@ namespace gz {
+  // _compoundShape which represents the collision.
+  // One exception is when the collision is a convex decomposed mesh.
+  // In this case, the child shape is another btCompoundShape (nested), and
+  // _childIndex is the index of one of the decomposed convex hulls
+  // in the nested compound shape. The nested compound shape is the collision.
+  int childCount = _compoundShape->getNumChildShapes();
+  if (childCount > 0)
+  {
+    if (_childIndex >= 0 && _childIndex < childCount)
+    {
+      // todo(iche033) We do not have sufficient info to determine which
+      // child shape is the collision if the link has convex decomposed mesh
+      // collisions alongside of other collisions. See following example:
+      // parentLink -> boxShape0
+      //            -> boxShape1
+      //            -> comopundShape -> convexShape0
+      //                             -> convexShape1
+      // A _childIndex of 1 is ambiguous as it could refer to either
+      // boxShape1 or convexShape1
+      // return nullptr in this case to indicate ambiguity
+      if (childCount > 1)
+      {
+        for (int i = 0; i < childCount; ++i)
+        {
+          const btCollisionShape *shape = _compoundShape->getChildShape(i);
+          if (this->HasConvexHullChildShapes(shape))
+            return nullptr;
+        }
+      }
+
+      const btCollisionShape *shape =
+          _compoundShape->getChildShape(_childIndex);
+      return shape;
+    }
+    else
+    {
+      return _compoundShape->getChildShape(0);
+    }
+  }
+  return nullptr;
+}
+
+/////////////////////////////////////////////////
+void GzCollisionDispatcher::dispatchAllCollisionPairs(
+    btOverlappingPairCache* pairCache,
+    const btDispatcherInfo& dispatchInfo,
+    btDispatcher* dispatcher)
+{
+  btCollisionDispatcher::dispatchAllCollisionPairs(
+      pairCache, dispatchInfo, dispatcher);
+
+  // Loop through all the contact manifolds.
+  // Find convex decomposed mesh collision shapes.
+  // Create a shared contact manifold for all decomposed shapes.
+  // This is so that we can limit the number of contact points to 4
+  // for the collision shape. Otherwise it will generate up to
+  // (decomposed col count * 4) contact points.
+  int numManifolds = this->getNumManifolds();
+  for (int i = 0; i < numManifolds; ++i)
+  {
+    btPersistentManifold* contactManifold =
+        this->getManifoldByIndexInternal(i);
+
+    const btMultiBodyLinkCollider* ob0 =
+        dynamic_cast<const btMultiBodyLinkCollider *>(
+        contactManifold->getBody0());
+    const btMultiBodyLinkCollider* ob1 =
+        dynamic_cast<const btMultiBodyLinkCollider *>(
+        contactManifold->getBody1());
+
+    // if it's a custom manifold, just refresh contacts, no need to
+    // loop through and check them.
+    if (this->customManifolds.find(contactManifold) !=
+        this->customManifolds.end())
+    {
+      contactManifold->refreshContactPoints(ob0->getWorldTransform(),
```

Comment:
Is this necessary? Do you know that Bullet itself won't be doing this?

Reply:
I remember I had to keep this call for the multi body link transform to be updated. As a test, I dropped a model with convex decomposed collisions on the ground -> this worked fine, but as soon as I moved it to a different pose, it's transforms are not updated.